airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -222,224 +212,356 @@ def _iter_upstream_conditions() -> Iterator[ColumnOperators]:
-                )
-            elif upstream_setup is None:  # for now, None only happens in mapped case
-                yield self._failing_status(
-                    reason=(
-                        f"Task's trigger rule '{trigger_rule}' cannot have mapped tasks as upstream. "
-                        f"upstream_states={upstream_states}, "
-                        f"upstream_task_ids={task.upstream_task_ids}"
+            elif trigger_rule == TR.ALL_DONE_SETUP_SUCCESS:
+                if not upstream_done:
+                    yield self._failing_status(
+                        reason=(
+                            f"Task's trigger rule '{trigger_rule}' requires all upstream tasks to have "
+                            f"completed, but found {len(upstream_tasks) - done} task(s) that were not done. "
+                            f"upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
                     )
-                )
-            elif upstream_setup and not success_setup >= 1:
-                yield self._failing_status(
-                    reason=(
-                        f"Task's trigger rule '{trigger_rule}' requires at least one upstream setup task be "
-                        f"successful, but found {upstream_setup - success_setup} task(s) that were not. "
-                        f"upstream_states={upstream_states}, "
-                        f"upstream_task_ids={task.upstream_task_ids}"
+                elif upstream_setup is None:  # for now, None only happens in mapped case
+                    yield self._failing_status(
+                        reason=(
+                            f"Task's trigger rule '{trigger_rule}' cannot have mapped tasks as upstream. "
+                            f"upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
+                    )
+                elif upstream_setup and not success_setup >= 1:
+                    yield self._failing_status(
+                        reason=(
+                            f"Task's trigger rule '{trigger_rule}' requires at least one upstream setup task "
+                            f"be successful, but found {upstream_setup - success_setup} task(s) that were "
+                            f"not. upstream_states={upstream_states}, "
+                            f"upstream_task_ids={task.upstream_task_ids}"
+                        )
                     )
-                )
-        else:
-            yield self._failing_status(reason=f"No strategy to evaluate trigger rule '{trigger_rule}'.")
+            else:
+                yield self._failing_status(reason=f"No strategy to evaluate trigger rule '{trigger_rule}'.")
+
+        # this dictionary enables use of task_id as the key for lru caching
+        # in function _get_relevant_upstream_map_indexes
+        tasks_dict = {}
```

Comment:
Probably want to move this above `_get_relevant_upstream_map_indexes` (or that function below this). There’s not much difference but better for readability and to prevent accidental access to this variable before it is created.

(Why do we want this dict to be empty when task is teadrown though? Or is there no difference?)

Reply:
see line 554, where we set the dict equal to `relevant_setups` when processing setup constraint.  then at 566 (if we get there) we reassign it to direct upstream tasks

Reply:
so, when `_get_relevant_upstream_map_indexes` is called for setup constraint, then it has access to the setup tasks, and when it is called for direct upstreams, it will have access to the direct upstreams.

we could avoid all of this by passing the task instead of the id, but for whatever reason it feels icky to me to do that since the hash functions are so weird, even if there's no practical way the object gets mutated during processing.  i can be easily convinced though.

Reply:
But `relevant_setups` is just a subset of all upstreams, isn’t it? Because I don’t see we catch KeyError anywhere, just passing the entire task dict shouldn’t change the logic at all from what I can tell. Is it just for performance?

Reply:
no, cus the bug we're solving here, is the case of _indirect_ setups -- setups that are not direct upstream relatives.  in normal operation this is a non issue.  but when you clear a work task, it clears its setups; if you don't add a check for the indirect setups then they will run concurrently. that's the problem here.

Reply:
separately, just a note, i expect there's a way to optimize this and reduce queries from 2 back down to 1; but i don't want to "change multiple things at once" in this way; i just want to layer the setup constraint on top of the existing trigger rule, and apply it only when dag has setups, to reduce potential for impact on dags not using the new feature, particularly given that here we're just trying to introduce a bugfix in a timely manner.

Reply:
No I mean, why can’t we just put all the tasks in that dict? Maybe `task_dict` should just be `ti.task.dag.task_dict`?

Reply:
yeah true, that should be fine.